File: 522. Longest Uncommon Subsequence II.cpp
/**
 * 
 * 思路，先按长度降序，从最长的字符串i开始，检查i是不是至少相同长度的字符串j的子序列，如果不是，就找到了
 * 
 * 
*/
class Solution {
public:
    int findLUSlength(vector<string>& strs) {
        int n = strs.size();
        sort(strs.begin(), strs.end(), [](const string &s1, const string &s2){
            return s1.size() > s2.size();
        });
        for (int i = 0; i < strs.size(); i++) {
            bool f = true;
            for (int j = 0; j < n && strs[j].size() >= strs[i].size(); j++) {
                if (i != j && isSubsequence(strs[i], strs[j])) {
                    f = false;
                    break;
                }
            }
            if (f) return strs[i].size();
        }
        return -1;
    }

    bool isSubsequence(string &a, string &b) {
        auto ia = a.begin();
        auto ib = b.begin();
        while (ia!=a.end() && ib!=b.end()) {
            if (*ia == *ib) 
                ia++;
            ib++;
        }
        return ia==a.end();
    }
};

File: go/src/leetcode/522/522. Longest Uncommon Subsequence II.go
/**
@author ZhengHao Lou
Date    2022/06/27
*/
package main

import (
	"fmt"
	"sort"
)

/**
https://leetcode.cn/problems/longest-uncommon-subsequence-ii/
*/
func findLUSlength(strs []string) int {
	n := len(strs)
	sort.Slice(strs, func(i, j int) bool {
		return len(strs[i]) > len(strs[j])
	})
NEXT:
	for i := range strs {
		for j := 0; j < n && len(strs[j]) >= len(strs[i]); j++ {
			if i != j && issub(strs[i], strs[j]) {
				continue NEXT
			}
		}
		return len(strs[i])
	}
	return -1
}

func issub(a, b string) bool {
	na, nb := len(a), len(b)
	var i, j int
	for i < na && j < nb {
		if a[i] == b[j] {
			i++
		}
		j++
	}
	return i == na
}

func main() {
	fmt.Println(findLUSlength([]string{"aba", "cdc", "eae"}))
	fmt.Println(findLUSlength([]string{"aaa", "aaa", "aa"}))
}
